woocommerce/woocommerce
# Move remote logger to `./src` and enhance `fetch_latest_woocommerce_version()` logic

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR addresses the feedback from the previous PR https://github.com/woocommerce/woocommerce/pull/49371.

Changes:

- Move remote logger class to `./src`
- Change `fetch_latest_woocommerce_version` transients to expire in 1 week
- Add retry timeout logic to `fetch_latest_woocommerce_version()` to prevent infinite fetch API when error occurs



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


1. Create a fresh install of WooCommerce
2. Install `Code Snippet` plugin
3. Add the following code snippet to your site:

```php
use Automattic\WooCommerce\Internal\Logging\RemoteLogger;

update_option( 'woocommerce_feature_remote_logging_enabled', 'yes' );

add_filter(
	'option_woocommerce_allow_tracking',
	function () {
		return 'yes';
	}
);

add_filter(
	'option_woocommerce_remote_variant_assignment',
	function () {
		return 5;
	}
);

$remote_logger = wc_get_container()->get( RemoteLogger::class );
$logging_allowed = $remote_logger->is_remote_logging_allowed();
$log_message = $logging_allowed ? 'Remote logging allowed' : 'Remote logging not allowed';
error_log($log_message);
```

4. Check the debug log for the message "Remote logging allowed".
5. Update code snippet to test the case when API request fails:

```php

use Automattic\WooCommerce\Internal\Logging\RemoteLogger;

update_option( 'woocommerce_feature_remote_logging_enabled', 'yes' );

add_filter(
	'option_woocommerce_allow_tracking',
	function () {
		return 'yes';
	}
);

add_filter(
	'option_woocommerce_remote_variant_assignment',
	function () {
		return 5;
	}
);

add_filter(
  'plugins_api', // phpcs:ignore PEAR.Functions.FunctionCallSignature.MultipleArguments
  function ( $result, $action, $args ) { // phpcs:ignore Generic.CodeAnalysis.UnusedFunctionParameter.FoundAfterLastUsed
    return new \WP_Error();
  },
  10,
  3
);

delete_transient( 'latest_woocommerce_version' );

$remote_logger = wc_get_container()->get( RemoteLogger::class );
$logging_allowed = $remote_logger->is_remote_logging_allowed();
$log_message = $logging_allowed ? 'Remote logging allowed' : 'Remote logging not allowed';
error_log($log_message);

$fetch_latest_woocommerce_version_retry = get_transient( 'fetch_latest_woocommerce_version_retry' );
error_log( 'fetch_latest_woocommerce_version_retry: ' . $fetch_latest_woocommerce_version_retry );
```

6. Refresh the page and check the debug log for the message "Remote logging not allowed" and the message "fetch_latest_woocommerce_version_retry: 1".
7. Refresh the page multiple times and check the debug log for the message "fetch_latest_woocommerce_version_retry: 2", "fetch_latest_woocommerce_version_retry: 3". It should not exceed 3 retries.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
